dropwizard-oauth/src/main/java/smartthings/dw/oauth/scope/ScopesAllowedDynamicFeature.java

```diff
@@ -12,60 +12,190 @@
 import javax.ws.rs.container.DynamicFeature;
 import javax.ws.rs.container.ResourceInfo;
 import javax.ws.rs.core.FeatureContext;
+import javax.ws.rs.core.MultivaluedMap;
 import java.io.IOException;
 import java.security.Principal;
+import java.util.*;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
 
 /**
  * Note that this code is based on {@link RolesAllowedDynamicFeature} but currently
  * does not intermix intelligently with those annotations. If a method or class uses
  * these annotations as well as the roll annotations both will trigger. Note that a
  * scope annotation on a method will override the annotation on a class.
+ *
+ * Filtering by scope can either be allowed by matching one of {@link ScopesAllowed} or
+ * {@link FineGrainedScopeAllowed}.
+ *
+ * {@link ScopesAllowed} provides matching against an exact string which must be available within
+ * the OAuth scopes.
+ *
+ * {@link FineGrainedScopeAllowed} provides dynamic matching using fine grained scopes format which can
+ * depend on values that are resolved through the request's context parameters.
  */
 public class ScopesAllowedDynamicFeature implements DynamicFeature {
+    // this pattern matches only camel case starting with lower case, so sorry if I am forcing
+    // you to conform, but this makes sense since they are for variables
+    private static final Pattern varNamePattern = Pattern.compile("\\{([a-z]\\w+)\\}");
 
 	@Override
 	public void configure(ResourceInfo resourceInfo, FeatureContext configuration) {
 		AnnotatedMethod am = new AnnotatedMethod(resourceInfo.getResourceMethod());
 
 		// ScopesAllowed on the method takes precedence over class level annotations
 		ScopesAllowed ra = am.getAnnotation(ScopesAllowed.class);
-		if (ra != null) {
-			configuration.register(new ScopesAllowedRequestFilter(ra.value()));
+		FineGrainedScopesAllowed fgra = am.getAnnotation(FineGrainedScopesAllowed.class);
+		if ((ra != null) || (fgra != null)) {
+			configuration.register(new ScopesAllowedRequestFilter(ra, fgra));
 			return;
 		}
 
 		ra = resourceInfo.getResourceClass().getAnnotation(ScopesAllowed.class);
-		if (ra != null) {
-			configuration.register(new ScopesAllowedRequestFilter(ra.value()));
+        fgra = resourceInfo.getResourceClass().getAnnotation(FineGrainedScopesAllowed.class);
+        if ((ra != null) || (fgra != null)) {
+			configuration.register(new ScopesAllowedRequestFilter(ra, fgra));
 		}
```

Comment:
Do we want to handle the case where `ScopesAllowed` is defined on the method, but `FineGrainedScopesAllowed` is defined on the Resource, or vice versa?  Or, is it assumed that any scope defined on the method supersedes everything defined on the resource?

As it stands now, if there is any scope defined on the method, then the resource level definitions will not be taken into account at all.

Reply:
You are right. I have fixed this issue, I think.